angr/knowledge_plugins/labels.py

```diff
@@ -1,66 +1,237 @@
+from collections import defaultdict, Iterable
+
+from bidict import bidict
+
 from .plugin import KnowledgeBasePlugin
+from ..misc.ux import deprecated
+
+
+class LabelsPlugin(KnowledgeBasePlugin):
+    """Storage for program labels. Access with kb.labels.    
+    
+    A label "name" is a label named "name", duh. Labels can be grouped into the seperate namespaces.
+    An address is allowed to have a different labels within different namespace, but it is not allowed
+    for address to have a different labels within a single namespace.
```

Comment:
This doesn't strike me as a good idea. Even within an elf file's symbol listing, the root from which we initially build the symbol database, there can be multiple names referring to a single address. You can see this sometimes in IDA, when it comments a line with "alternate names: xxx, yyy, zzz"

I would appreciate being able to have this feature where there's a one-to-one mapping from address to name within a namespace, but it doesn't reflect the reality we have to work with. Potential fixes:

- remove this requirement, and have `get_name` and `get_all_names` functions
- add an "alternate names" section to the plugin to handle these

Reply:
Personally, I'm for not removing this requirement. Aren't the "alternate names" in IDA just weak refs or is there something else? If they're weak refs, that's an attribute of the `Symbol` so it should be separate from the label (i.e., the user can set the label for a symbol to the alternate name if they choose to), either the separate "alternate names" thing you mentioned, or even a separate namespace. With "alternate names" a `get_all_names` function for the plugin could return the combination of label names and alternate names.

Reply:
If by "weak refs" you mean weak symbols, then no, you can have just normal symbols occupying the same address:

```asm
.intel_syntax noprefix

.global _start
.global _start2
_start:
_start2:
mov rax, 60
mov rdi, 42
syscall
```

```shellsession
rhelmot@rhelmot-VirtualBox:~$ as test.s -o test.o
rhelmot@rhelmot-VirtualBox:~$ ld test.o -o test
rhelmot@rhelmot-VirtualBox:~$ readelf -s test

Symbol table '.symtab' contains 7 entries:
   Num:    Value          Size Type    Bind   Vis      Ndx Name
     0: 0000000000000000     0 NOTYPE  LOCAL  DEFAULT  UND
     1: 0000000000400078     0 SECTION LOCAL  DEFAULT    1
     2: 0000000000400078     0 NOTYPE  GLOBAL DEFAULT    1 _start2
     3: 0000000000400078     0 NOTYPE  GLOBAL DEFAULT    1 _start
     4: 0000000000601000     0 NOTYPE  GLOBAL DEFAULT    1 __bss_start
     5: 0000000000601000     0 NOTYPE  GLOBAL DEFAULT    1 _edata
     6: 0000000000601000     0 NOTYPE  GLOBAL DEFAULT    1 _end
```